### PR TITLE
Fix #13598 by protecting partial_pred

### LINKED ISSUE #13598: Falsely(?) triggered warning 56 [unreachable-case] starting with 5.3.0~beta1
Trying out 5.3.0\~beta1 with Unison, I see the following warning. This warning is not present with versions up to and including 5.3.0\~alpha1.

```
ocamlopt -g -I lwt -I ubase -I system -I +unix -I +str -I system/generic -I lwt/generic -c copy.ml
File "/home/runner/work/unison/unison/src/copy.ml", line 218, characters 37-88:
218 |                                    | TransferNeedsDoubleCheckAgainstCurrentSource (a, b) -> I32 (a, b)
                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Warning 56 [unreachable-case]: this match case is unreachable.
Consider replacing it with a refutation case '<pat> -> .'
```

The source: https://github.com/bcpierce00/unison/blob/96e15a800bb0e619da6f7092ddefc47101c3092f/src/copy.ml#L202-L224

### PR DESCRIPTION
**Pull request by garrigue**
This is an alternative fix to #13598, this time protecting `partial_pred` with `trace_gadt_instances ~force:true`, to ensure that we do not look up old abbreviations. (That was a trick one.)

It is required, since the behaviour of the exhaustivity check was potentially buggy, but does not preclude merging #13599 too.

### CODE REVIEW

File: typing/typecore.ml [hunk 1]
```diff
@@ -2533,7 +2533,8 @@ let partial_pred ~lev ~splitting_mode ?(explode=0) env expected_ty p =
       } in
   try
     let typed_p =
-      check_counter_example_pat ~counter_example_args penv p expected_ty
+      wrap_trace_gadt_instances ~force:true !!penv
+        (check_counter_example_pat ~counter_example_args penv p) expected_ty
```
Comment (Octachron) [inline typing/typecore.ml]: Nitpick: since this is the only call to `check_counter_example_pat`, and calling the function without flushing the abbreviation cache is wrong, I propose to move the wrapping to the definition of `check_counter_example_pat`.

File: typing/ctype.ml [hunk 2]
```diff
@@ -1578,6 +1577,7 @@ let check_abbrev_env env =
   if not (Env.same_type_declarations env !previous_env) then begin
     (* prerr_endline "cleanup expansion cache"; *)
     cleanup_abbrev ();
+    simple_abbrevs := Mnil;
```
Comment (Octachron) [inline typing/ctype.ml:1580]: This is an orthogonal change, isn't it? When `trace_gadt_instances` is on, we always use `abbrevs` and not `simple_abbrevs`. At the very least, the change has no effect on the testsuite.

**Review by Octachron — APPROVED** [review]
I agree that flushing the abbreviation cache is currently needed when checking counter-examples to make sure that we don't reuse old expansions at a lower level that makes `reify` fails.

**Comment by Octachron** [comment]
@garrigue , would you prefer that we merge this PR in its current state while I propose my minor changes as PR? I am keen to have this fix in the beta2 for OCaml 5.3 .

File: typing/ctype.ml (hunk 2 above)
Comment (garrigue) [reply typing/ctype.ml:1580]: It is orthogonal indeed (i.e., it does not apply to this specific case), but I have been always wondering why this caused no bug, and could never find something else to put it in. Since we are fixing bugs about cleaning abbrevs, let's do it properly.

**Comment by garrigue** [comment]
I have rebased and followed your suggestion.
I also add the counterexample from this morning, since we're not going to make a separate PR just for it.

**Comment by garrigue** [comment]
Also, since we cannot merge #13599, I think we need #13585.

**Comment by Octachron** [comment]
Yes, I agree that we want to merge #13585 to reduce the change of behavior with 5.2 .

**Comment by Octachron** [comment]
Cherry-picked on 5.3 as e20cc199460f3c .
